tools/config_config.py

```diff
@@ -589,100 +598,22 @@ def translation_configs(opts):
-                stack['adapters'] = adapters
-        key = str(tgt_stack)    # An ugly way to freeze the mutable structure
-        decoder_stacks[tgt_lang][key] = tgt_stack
-        # Transforms and subword models also need to be respecified during translation
-        if 'transforms' not in task_opts:
-            transforms = None
-        else:
-            transforms = [
-                transform for transform in task_opts['transforms']
-                if not transform == 'filtertoolong'
-            ]
-        transforms_by_lang[src_lang] = transforms
-        # Write config for the supervised directions
-        _write_translation_config(
-            src_lang,
-            tgt_lang,
-            src_stack,
-            tgt_stack,
-            transforms,
-            src_subword_model,
-            tgt_subword_model,
-            'supervised',
-            translation_config_dir,
-            use_src_lang_token,
-        )
-        supervised_pairs.add((src_lang, tgt_lang))
-    if zero_shot:
-        src_langs = encoder_stacks.keys()
-        tgt_langs = decoder_stacks.keys()
-        # verify that there is an unique stack for each language
-        ambiguous_src = [src_lang for src_lang in encoder_stacks if len(encoder_stacks[src_lang]) > 1]
-        ambiguous_tgt = [tgt_lang for tgt_lang in decoder_stacks if len(decoder_stacks[tgt_lang]) > 1]
-        if len(ambiguous_src) > 0 or len(ambiguous_tgt) > 0:
-            raise Exception(
-                'Zero-shot translation configs can only be generated if each source (target) language '
-                'has an unambigous encoder (decoder) stack.\n'
-                'The following languages have more than one encoder/decoder stack:\n'
-                f'Source: {ambiguous_src}\nTarget: {ambiguous_tgt}'
-            )
-        for src_lang in src_langs:
-            for tgt_lang in tgt_langs:
-                if src_lang == tgt_lang:
-                    continue
-                if (src_lang, tgt_lang) in supervised_pairs:
-                    continue
-                src_stack = list(encoder_stacks[src_lang].values())[0]
-                tgt_stack = list(decoder_stacks[tgt_lang].values())[0]
-                transforms = transforms_by_lang[src_lang]
-                _write_translation_config(
-                    src_lang,
-                    tgt_lang,
-                    src_stack,
-                    tgt_stack,
-                    transforms,
-                    src_subword_model,
-                    tgt_subword_model,
-                    'zeroshot',
-                    translation_config_dir,
-                    use_src_lang_token,
-                )
+    # src_subword_model = opts.in_config[0].get('src_subword_model', None)
+    # tgt_subword_model = opts.in_config[0].get('tgt_subword_model', None)
+    # use_src_lang_token = cc_opts.get('use_src_lang_token', False)
+
+    # TODO: create zero-shot tasks using the same template as for the supervised tasks, except that:
```

Comment:
to be discussed. There's a potential risk for confusion by having two very similar types of tasks that need to be configured in slightly different ways.
I think we might want either of the following to:
1.  explicitly mark some tasks as test-only (i.e., zeroshot), 
2. halt and catch fire in partially / improperly defined tasks (any defined `path_src`  should require device assignment and explode mid-air otherwise).

We might want a big fat DEBUG message about which tasks are defined for which steps (train/val/test), e.g. a tab-spaced table with split as columns and tasks as rows.

Reply:
Yes, if we generate the zero-shots already in the training config (i.e. before training), then we need to clearly mark the test-only tasks and add validation to prevent accidental mistakes.

However, I see this more as an extra transformation to run on a training config to add the zero-shot tasks, so that you can then easily use it as a translation config. You shouldn't need to decide before training whether you want a particular zero-shot or not.